src/Razor/Microsoft.AspNetCore.Mvc.Razor.Extensions/src/ViewComponentTagHelperDescriptorProvider.cs

```diff
@@ -37,15 +37,18 @@ public void Execute(TagHelperDescriptorProviderContext context)
 
             var types = new List<INamedTypeSymbol>();
             var visitor = new ViewComponentTypeVisitor(vcAttribute, nonVCAttribute, types);
-            var discoveryMode = context.Items.GetTagHelperDiscoveryFilter();
 
-            if ((discoveryMode & TagHelperDiscoveryFilter.CurrentCompilation) == TagHelperDiscoveryFilter.CurrentCompilation)
+            var targetReference = context.Items.GetTargetMetadataReference();
+            if (targetReference is MetadataReference && compilation.GetAssemblyOrModuleSymbol(targetReference) is IAssemblySymbol targetAssembly)
```

Comment:
it might not super matter, but it might help to move the `&&` check in to the if. It would be weird if this fell back to discovering everything just because it could not turn it into an IAssemblySymbol

Something like this:
```suggestion
            if (targetReference is not null)
            {
               if (compilation.GetAssemblyOrModuleSymbol(targetReference) is IAssemblySymbol targetAssembly)
               { 
                   ...
             
             
```